### PR TITLE
leger-api: add GenMap support

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances #2256.

It adds support of GenMap to the ledger API.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
Should I add a line in the release notes ? 
Add a new data in the protobuf, but can be used only with the development version of LF.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/ValueValidator.scala [hunk 1]
```diff
@@ -100,28 +91,42 @@ object ValueValidator {
...
+    case Sum.GenMap(genMap0) =>
+      val genMap = genMap0.entries
+        .foldLeft[Either[StatusRuntimeException, BackStack[(domain.Value, domain.Value)]]](
+          Right(BackStack.empty)) {
+          case (acc, api.GenMap.Entry(key0, value0)) =>
+            for {
+              stack <- acc
+              key <- requirePresence(key0, "key")
+              validatedKey <- validateValue(key)
+              value <- requirePresence(value0, "value")
```
Comment (gerolf-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/ValueValidator.scala]: We can potentially avoid some work by checking for the presence of the `value` field before validating the `key` value.

Comment (remyhaemmerle-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/ValueValidator.scala]: I inversed line 118 and 119.

**Comment by associahedron** [comment]
> Should I add a line in the release notes ?
> Add a new data in the protobuf, but can be used only with the development version of LF.

IMO it's best to leave `GenMap` off the release notes until `GenMap` becomes stabilized.

**Comment by remyhaemmerle-da** [comment]
@associahedron 
> IMO it's best to leave `GenMap` off the release notes until `GenMap` becomes stabilized.

I would like, but this change will break scala/haskell code (with a non exhaustive pattern matching) that relies on the code generated from the proto definition.

**Pull request merged**